modules/images/image-loading-optimization/class-ilo-html-tag-processor.php

```diff
@@ -0,0 +1,308 @@
+<?php
+/**
+ * Image Loading Optimization: ILO_HTML_Tag_Processor class
+ *
+ * @package performance-lab
+ * @since n.e.x.t
+ */
+
+/**
+ * Processor leveraging WP_HTML_Tag_Processor which walks over a document and gathers breadcrumbs and invokes a callback for each open tag.
+ *
+ * @since n.e.x.t
+ * @access private
+ */
+final class ILO_HTML_Tag_Processor {
+
+	/**
+	 * HTML elements that are self-closing.
+	 *
+	 * @link https://www.w3.org/TR/html5/syntax.html#serializing-html-fragments
+	 * @link https://github.com/ampproject/amp-toolbox-php/blob/c79a0fe558a3c042aee4789bbf33376cca7a733d/src/Html/Tag.php#L206-L232
+	 *
+	 * @var string[]
+	 */
+	const SELF_CLOSING_TAGS = array(
+		'AREA',
+		'BASE',
+		'BASEFONT',
+		'BGSOUND',
+		'BR',
+		'COL',
+		'EMBED',
+		'FRAME',
+		'HR',
+		'IMG',
+		'INPUT',
+		'KEYGEN',
+		'LINK',
+		'META',
+		'PARAM',
+		'SOURCE',
+		'TRACK',
+		'WBR',
+	);
+
+	/**
+	 * The set of HTML tags whose presence will implicitly close a <p> element.
+	 * For example '<p>foo<h1>bar</h1>' should parse the same as '<p>foo</p><h1>bar</h1>'.
+	 *
+	 * @link https://www.w3.org/TR/html-markup/p.html
+	 * @link https://github.com/ampproject/amp-toolbox-php/blob/c79a0fe558a3c042aee4789bbf33376cca7a733d/src/Html/Tag.php#L262-L293
+	 *
+	 * @var string[]
+	 */
+	const P_CLOSING_TAGS = array(
+		'ADDRESS',
+		'ARTICLE',
+		'ASIDE',
+		'BLOCKQUOTE',
+		'DIR',
+		'DL',
+		'FIELDSET',
+		'FOOTER',
+		'FORM',
+		'H1',
+		'H2',
+		'H3',
+		'H4',
+		'H5',
+		'H6',
+		'HEADER',
+		'HR',
+		'MENU',
+		'NAV',
+		'OL',
+		'P',
+		'PRE',
+		'SECTION',
+		'TABLE',
+		'UL',
+	);
+
+	/**
+	 * Open stack tags.
+	 *
+	 * @var string[]
+	 */
+	private $open_stack_tags = array();
+
+	/**
+	 * Open stag indices.
+	 *
+	 * @var int[]
+	 */
+	private $open_stack_indices = array();
+
+	/**
+	 * Processor.
+	 *
+	 * @var WP_HTML_Tag_Processor
+	 */
+	private $processor;
+
+	/**
+	 * Constructor.
+	 *
+	 * @param string $html HTML to process.
+	 */
+	public function __construct( string $html ) {
+		$this->processor = new WP_HTML_Tag_Processor( $html );
+	}
+
+	/**
+	 * Walk over the document.
+	 *
+	 * Whenever an open tag is encountered, invoke the supplied $open_tag_callback and pass the tag name and breadcrumbs.
+	 *
+	 * @param callable $open_tag_callback Open tag callback. The processor instance is passed as the sole argument.
+	 */
+	public function walk( callable $open_tag_callback ) {
```

Comment:
Actually, I think I'll reimplement this as a generator and use the `next_tag` method name for additional parity with `WP_HTML_Tag_Processor`